### PR TITLE
Make Qna log TraceActivity

### PR DESCRIPTION
**Pull request by daveta**
Make Qna log TraceActivity events.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 1]
```diff
@@ -84,18 +90,19 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
...
-        public async Task<QueryResult[]> GetAnswersAsync(string question)
+        public async Task<QueryResult[]> GetAnswersAsync(ITurnContext context)
         {
             var requestUrl = $"{_endpoint.Host}/knowledgebases/{_endpoint.KnowledgeBaseId}/generateanswer";
 
             var request = new HttpRequestMessage(HttpMethod.Post, requestUrl);
+            var messageActivity = context?.Activity?.AsMessageActivity();
 
             var jsonRequest = JsonConvert.SerializeObject(
                 new
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Do we want to leave this as an Anonymous type? Doing so means we have no tests / regressions around it. I would much prefer this to be a strong type with actual Unit tests.

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 2]
```diff
@@ -84,18 +90,19 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
...
                 {
-                    question,
+                    messageActivity?.Text,
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Is there a scenario here where the text is empty (or the activity is null), and we would want to log?

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 3]
```diff
@@ -129,7 +136,22 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
                     answer.Score = answer.Score / 100;
                 }
 
-                return results.Answers.Where(answer => answer.Score > _options.ScoreThreshold).ToArray();
+                var result = results.Answers.Where(answer => answer.Score > _options.ScoreThreshold).ToArray();
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Is this ToArray() needed here?

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 4]
```diff
@@ -129,7 +136,22 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
...
+                    StrictFilters = _options.StrictFilters,
+                    MetadataBoost = _options.MetadataBoost,
+                };
+                var traceActivity = Activity.CreateTraceActivity(QnAMakerMiddlewareName, QnAMakerTraceType, traceInfo, QnAMakerTraceLabel);
+                await context.SendActivityAsync(traceActivity).ConfigureAwait(false);
+
+                return result;
             }
 
             return null;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Can this code be refactored so the "main" case of the method doesn't return null? That just seems weird.

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 5]
```diff
@@ -84,18 +90,19 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
         /// <summary>
         /// Generates an answer from the knowledge base.
         /// </summary>
-        /// <param name="question">The user question to be queried against your knowledge base.</param>
+        /// <param name="context">The Turn Context that contains the user question to be queried against your knowledge base.</param>
         /// <returns>A list of answers for the user query, sorted in decreasing order of ranking score.</returns>
-        public async Task<QueryResult[]> GetAnswersAsync(string question)
+        public async Task<QueryResult[]> GetAnswersAsync(ITurnContext context)
         {
```
[the remark this answers is not in the document]
Comment (daveta) [reply libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: fixed

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs [hunk 6]
```diff
@@ -84,18 +90,19 @@ public QnAMaker(QnAMakerEndpoint endpoint, QnAMakerOptions options = null, HttpC
...
             var requestUrl = $"{_endpoint.Host}/knowledgebases/{_endpoint.KnowledgeBaseId}/generateanswer";
 
             var request = new HttpRequestMessage(HttpMethod.Post, requestUrl);
+            var messageActivity = context?.Activity?.AsMessageActivity();
```
[the remark this answers is not in the document]
Comment (daveta) [reply libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: yeah, suspiciously there was no check before, but can't think of any reason why a null message makes sense.  I'll reject on both a non-message and empty string..

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs (hunk 2 above)
Comment (daveta) [reply libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Yeah, assuming there isn't - i'll block having empty or null text as well.

File: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs (hunk 4 above)
Comment (daveta) [reply libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs]: Heh, good idea.

**Comment by cleemullins** [comment]
@daveta, There is a build error. The error is:

tests\Microsoft.Bot.Builder.Ai.QnA.Tests\QnAMakerTests.cs(53,41): Error CS1593: Delegate 'BotCallbackHandler' does not take 1 arguments

Process 'msbuild.exe' exited with code '1'.

I think that's probably a Rebase error with the work John did yesterday around CancellationTokens.

**Comment by cleemullins** [comment]
>             }

Have I mentioned how much I dislike this single line of code? Comparing scores feels like a mistake here. A problem for another day. 

---
Refers to: libraries/Microsoft.Bot.Builder.Ai.QnA/QnAMaker.cs:62 in 4209b70. [](commit_id = 4209b705836f864516928214ee3fc8cb6bd90c2a, deletion_comment = False)

**Comment by cleemullins** [comment]
@daveta. The new LUIS Recognizer Tests are failing now:

```
2018-07-26T14:24:37.5518323Z  Assert.Fail failed. Returned JSON in ..\..\..\TestData\Composite1.json.new != expected JSON in ..\..\..\TestData\Composite1.json
2018-07-26T14:24:37.5519121Z Stack Trace:
2018-07-26T14:24:37.5519567Z    at Microsoft.Bot.Builder.Ai.Luis.Tests.LuisRecognizerTests.<TestJson>d__10`1.MoveNext() in D:\a\1\s\tests\Microsoft.Bot.Builder.Ai.LUIS.Tests\LuisRecognizerTests.cs:line 273
2018-07-26T14:24:37.5519978Z --- End of stack trace from previous location where exception was thrown ---
2018-07-26T14:24:37.5520308Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-26T14:24:37.5520674Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-26T14:24:37.5521110Z    at Microsoft.Bot.Builder.Ai.Luis.Tests.LuisRecognizerTests.<Composite1>d__12.MoveNext() in D:\a\1\s\tests\Microsoft.Bot.Builder.Ai.LUIS.Tests\LuisRecognizerTests.cs:line 330
2018-07-26T14:24:37.5521508Z --- End of stack trace from previous location where exception was thrown ---
2018-07-26T14:24:37.5521913Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-26T14:24:37.5522512Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Pull request merged**